cou/apps/app.py

```diff
@@ -245,24 +224,121 @@ def __str__(self) -> str:
+                f"compatible OpenStack release."
+            ) from exc
+
+    @property
+    def channel(self) -> str:
+        """Get charm channel of the application.
+
+        :return: Charm channel. E.g: ussuri/stable
+        :rtype: str
+        """
+        return self._channel
+
+    @channel.setter
+    def channel(self, charm_channel: str) -> None:
+        """Set charm channel of the application.
+
+        When application comes from charm store, the channel won't be OpenStack related.
+        :param charm_channel: Charm channel. E.g: ussuri/stable
+        :type charm_channel: str
+        :raises ValueError:  Exception raised when channel is not a valid OpenStack
+            channel.
+        """
+        if (
+            self.is_from_charm_store
+            or (not self.is_os_channel_based and self.is_subordinate)
+            or self.is_valid_track
+        ):
+            self._channel = charm_channel
+            return
+        raise ValueError(self.channel_err_msg(charm_channel))
+
+    def channel_err_msg(self, charm_channel: str) -> str:
+        """Error message when channel is not valid.
+
+        :param charm_channel: Charm channel.
+        :type charm_channel: str
+        :return: error message
+        :rtype: str
+        """
+        return f"'{self.name}' has invalid channel: '{charm_channel}'"
+
     @property
     def possible_current_channels(self) -> list[str]:
         """Return the possible current channels based on the current OpenStack release.
 
         :return: The possible current channels for the application. E.g: ["ussuri/stable"]
         :rtype: list[str]
         """
-        return [f"{self.current_os_release.codename}/stable"]
+        return self._get_channels_based_on_os(self.current_os_release)
 
     def target_channel(self, target: OpenStackRelease) -> str:
         """Return the appropriate channel for the passed OpenStack target.
 
         :param target: OpenStack release as target to upgrade.
         :type target: OpenStackRelease
+        :raises ApplicationError: When cannot find a suitable target channel.
         :return: The next channel for the application. E.g: victoria/stable
         :rtype: str
         """
-        return f"{target.codename}/stable"
+        targets_channels = self._get_channels_based_on_os(target)
+        if targets_channels:
+            return targets_channels[-1]
+        raise ApplicationError(
+            f"Cannot find a suitable '{self.charm}' charm channel for {target.codename}"
+        )
+
+    def _get_channels_based_on_os(self, os_release: OpenStackRelease) -> list[str]:
+        """Get the channels based on the OpenStack release.
+
+        :param os_release: OpenStack release.
+        :type os_release: OpenStackRelease
+        :return: List of possible channels compatible with a OpenStack release.
+        :rtype: list[str]
+        """
+        return [f"{os_release.codename}/stable"]
```

Comment:
Again this function makes just more room to error and does not same any time to use it, check my previous comment.

Reply:
I don't agree. This function we are able to share `target_channel` and `possible_current_channels` on `OpenStackAuxiliaryApplication` and `OpenStackApplication`. It's one simple function that save us overwriting two other functions

Reply:
calling `f"{os_release.codename}/stable"` is easier than calling `_get_channels_based_on_os`